File: fons/norma-ts/yaml.ts
/**
 * yaml.ts - YAML Encoding/Decoding Implementation
 *
 * Native TypeScript implementation of the HAL YAML interface.
 * Uses Bun's built-in YAML support (globalThis.Bun.YAML or js-yaml fallback).
 *
 * Verb meanings:
 *   - pange (compose): serialize value to YAML string
 *   - necto (bind): bind multiple documents into multi-doc YAML
 *   - solve (untangle): parse YAML string to value
 *   - tempta (try): attempt to parse, return null on error
 *   - collige (gather): gather all documents from multi-doc YAML
 */

// Bun has built-in YAML support
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const YAML = (globalThis as any).Bun?.YAML ?? await import('js-yaml').then(m => ({
    parse: m.load,
    stringify: m.dump,
}));

export const yaml = {
    // =========================================================================
    // SERIALIZATION
    // =========================================================================

    /** Serialize value to YAML string */
    pange(valor: unknown): string {
        return YAML.stringify(valor);
    },

    /** Bind multiple documents into multi-doc YAML string */
    necto(documenta: unknown[]): string {
        return documenta.map(doc => YAML.stringify(doc)).join('---\n');
    },

    // =========================================================================
    // PARSING
    // =========================================================================

    /** Parse YAML string to value (throws on error) */
    solve(yaml: string): unknown {
        return YAML.parse(yaml);
    },

    /** Attempt to parse YAML string (returns null on error) */
    tempta(yaml: string): unknown | null {
        try {
            return YAML.parse(yaml);
        }
        catch {
            return null;
        }
    },

    /** Gather all documents from multi-doc YAML string */
    collige(yaml: string): unknown[] {
        const docs = yaml.split(/^---\s*$/m).filter(s => s.trim());
        return docs.map(doc => YAML.parse(doc));
    },

    // =========================================================================
    // TYPE CHECKING
    // =========================================================================

    estNihil(valor: unknown): boolean {
        return valor === null || valor === undefined;
    },

    estBivalens(valor: unknown): boolean {
        return typeof valor === 'boolean';
    },

    estNumerus(valor: unknown): boolean {
        return typeof valor === 'number';
    },

    estTextus(valor: unknown): boolean {
        return typeof valor === 'string';
    },

    estLista(valor: unknown): boolean {
        return Array.isArray(valor);
    },

    estTabula(valor: unknown): boolean {
        return typeof valor === 'object' && valor !== null && !Array.isArray(valor);
    },

    // =========================================================================
    // VALUE EXTRACTION
    // =========================================================================

    utTextus(valor: unknown, defVal: string): string {
        return typeof valor === 'string' ? valor : defVal;
    },

    utNumerus(valor: unknown, defVal: number): number {
        return typeof valor === 'number' ? valor : defVal;
    },

    utBivalens(valor: unknown, defVal: boolean): boolean {
        return typeof valor === 'boolean' ? valor : defVal;
    },

    // =========================================================================
    // VALUE ACCESS
    // =========================================================================

    /** Get value by key (returns null if missing) */
    cape(valor: unknown, clavis: string): unknown {
        if (typeof valor === 'object' && valor !== null && !Array.isArray(valor)) {
            return (valor as Record<string, unknown>)[clavis] ?? null;
        }
        return null;
    },

    /** Pluck value by array index (returns null if out of bounds) */
    carpe(valor: unknown, index: number): unknown {
        if (Array.isArray(valor) && index >= 0 && index < valor.length) {
            return valor[index];
        }
        return null;
    },

    /** Find value by dotted path (returns null if not found) */
    inveni(valor: unknown, via: string): unknown {
        const parts = via.split('.');
        let current: unknown = valor;

        for (const part of parts) {
            if (current === null || current === undefined) {
                return null;
            }
            if (typeof current === 'object' && !Array.isArray(current)) {
                current = (current as Record<string, unknown>)[part];
            }
            else if (Array.isArray(current)) {
                const idx = parseInt(part, 10);
                if (isNaN(idx) || idx < 0 || idx >= current.length) {
                    return null;
                }
                current = current[idx];
            }
            else {
                return null;
            }
        }

        return current ?? null;
    },
};

File: runtimes/norma-go/yaml/yaml.go
// Package yaml provides YAML encoding/decoding for the Faber HAL.
//
// Etymology: "YAML" - YAML Ain't Markup Language (universal, kept as-is).
package yaml

import (
	"reflect"
	"strings"

	"gopkg.in/yaml.v3"
)

// =============================================================================
// SERIALIZATION
// =============================================================================

// Pange serializes value to YAML string.
func Pange(valor any) (string, error) {
	data, err := yaml.Marshal(valor)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// Necto binds multiple documents into multi-doc YAML string.
func Necto(documenta []any) (string, error) {
	var parts []string
	for _, doc := range documenta {
		data, err := yaml.Marshal(doc)
		if err != nil {
			return "", err
		}
		parts = append(parts, string(data))
	}
	return strings.Join(parts, "---\n"), nil
}

// =============================================================================
// PARSING
// =============================================================================

// Solve parses YAML string to value.
func Solve(yamlStr string) (any, error) {
	var result any
	err := yaml.Unmarshal([]byte(yamlStr), &result)
	if err != nil {
		return nil, err
	}
	return normalizeYAML(result), nil
}

// Tempta attempts to parse YAML string, returns nil on error.
func Tempta(yamlStr string) any {
	result, err := Solve(yamlStr)
	if err != nil {
		return nil
	}
	return result
}

// Collige gathers all documents from multi-doc YAML string.
func Collige(yamlStr string) ([]any, error) {
	decoder := yaml.NewDecoder(strings.NewReader(yamlStr))
	var results []any

	for {
		var doc any
		err := decoder.Decode(&doc)
		if err != nil {
			if err.Error() == "EOF" {
				break
			}
			return nil, err
		}
		results = append(results, normalizeYAML(doc))
	}

	return results, nil
}

// =============================================================================
// TYPE CHECKING
// =============================================================================

// EstNihil checks if value is nil.
func EstNihil(valor any) bool {
	return valor == nil
}

// EstBivalens checks if value is a boolean.
func EstBivalens(valor any) bool {
	_, ok := valor.(bool)
	return ok
}

// EstNumerus checks if value is a number.
func EstNumerus(valor any) bool {
	switch valor.(type) {
	case float64, float32, int, int64, int32, int16, int8, uint, uint64, uint32, uint16, uint8:
		return true
	default:
		return false
	}
}

// EstTextus checks if value is a string.
func EstTextus(valor any) bool {
	_, ok := valor.(string)
	return ok
}

// EstLista checks if value is an array/slice.
func EstLista(valor any) bool {
	if valor == nil {
		return false
	}
	v := reflect.ValueOf(valor)
	return v.Kind() == reflect.Slice || v.Kind() == reflect.Array
}

// EstTabula checks if value is an object/map.
func EstTabula(valor any) bool {
	if valor == nil {
		return false
	}
	v := reflect.ValueOf(valor)
	return v.Kind() == reflect.Map
}

// =============================================================================
// VALUE EXTRACTION
// =============================================================================

// UtTextus extracts string value or returns default.
func UtTextus(valor any, defVal string) string {
	if s, ok := valor.(string); ok {
		return s
	}
	return defVal
}

// UtNumerus extracts numeric value or returns default.
func UtNumerus(valor any, defVal int) int {
	switch v := valor.(type) {
	case float64:
		return int(v)
	case float32:
		return int(v)
	case int:
		return v
	case int64:
		return int(v)
	case int32:
		return int(v)
	default:
		return defVal
	}
}

// UtBivalens extracts boolean value or returns default.
func UtBivalens(valor any, defVal bool) bool {
	if b, ok := valor.(bool); ok {
		return b
	}
	return defVal
}

// =============================================================================
// VALUE ACCESS
// =============================================================================

// Cape gets value by key from object (returns nil if missing).
func Cape(valor any, clavis string) any {
	if m, ok := valor.(map[string]any); ok {
		return m[clavis]
	}
	return nil
}

// Carpe plucks value by index from array (returns nil if out of bounds).
func Carpe(valor any, index int) any {
	if arr, ok := valor.([]any); ok {
		if index >= 0 && index < len(arr) {
			return arr[index]
		}
	}
	return nil
}

// Inveni finds value by dotted path (returns nil if not found).
func Inveni(valor any, via string) any {
	parts := strings.Split(via, ".")
	current := valor

	for _, part := range parts {
		if current == nil {
			return nil
		}
		if m, ok := current.(map[string]any); ok {
			current = m[part]
		} else {
			return nil
		}
	}

	return current
}

// =============================================================================
// HELPERS
// =============================================================================

// normalizeYAML converts map[any]any to map[string]any for consistency.
// YAML v3 can produce map[string]any but we normalize to be safe.
func normalizeYAML(v any) any {
	switch val := v.(type) {
	case map[string]any:
		result := make(map[string]any)
		for k, v := range val {
			result[k] = normalizeYAML(v)
		}
		return result
	case map[any]any:
		result := make(map[string]any)
		for k, v := range val {
			if ks, ok := k.(string); ok {
				result[ks] = normalizeYAML(v)
			}
		}
		return result
	case []any:
		result := make([]any, len(val))
		for i, v := range val {
			result[i] = normalizeYAML(v)
		}
		return result
	default:
		return v
	}
}
